tests/TwoFactor/LoginFormTest.php

```diff
@@ -0,0 +1,175 @@
+            ]
+        );
+        $this->assertFalse($form->validate(['password']));
+    }
+
+    public function testInvalidLogin(): void
+    {
+        $form = new Authentication\TwoFactor\LoginForm(
+            $this->createMock(Http\Request::class),
+            $this->createMock(Http\Response::class),
+            [
+                'repository' => $this->createMock(Token\Repository::class),
+                'tokenGenerator' => $this->createMock(Mock\TokenGenerator::class),
+            ]
+        );
+        $form->login = 'test';
+        $form->password = 'test';
+
+        $form->identityClass = Mock\Identity::class;
+        Mock\Identity::$willFind = null;
+
+        try {
+            $form->getResponse();
+        } catch (Http\Exceptions\HttpValidationException $exception) {
+            $this->assertEquals(
+                'login is invalid.',
+                $form->getFirstError('login')
+            );
+        }
+        $this->assertInstanceOf(Http\Exceptions\HttpValidationException::class, $exception);
+    }
+
+    public function testInvalidPassword(): void
+    {
+        $form = new Authentication\TwoFactor\LoginForm(
+            $this->createMock(Http\Request::class),
+            $this->createMock(Http\Response::class),
+            [
+                'repository' => $this->createMock(Token\Repository::class),
+                'tokenGenerator' => $this->createMock(Mock\TokenGenerator::class),
+            ]
+        );
+        $form->login = 'test';
+        $form->password = 'test';
+
+        $form->identityClass = Mock\Identity::class;
+        Mock\Identity::$willFind = $identity = $this->createMock(Mock\Identity::class);
+
+        $identity->expects($this->once())
+            ->method('validatePassword')
+            ->with('test')
+            ->willReturn(false);
+
+        try {
+            $form->getResponse();
+        } catch (Http\Exceptions\HttpValidationException $exception) {
+            $this->assertEquals(
+                'password is invalid.',
+                $form->getFirstError('password')
+            );
+        }
+        $this->assertInstanceOf(Http\Exceptions\HttpValidationException::class, $exception);
+    }
+
+    public function testSuccess(): void
+    {
+        $form = new Authentication\TwoFactor\LoginForm(
+            $this->createMock(Http\Request::class),
+            $this->createMock(Http\Response::class),
+            [
+                'repository' => $repository = $this->createMock(Token\Repository::class),
+                'tokenGenerator' => $tokenGenerator = $this->createMock(Mock\TokenGenerator::class),
+            ]
+        );
+        $form->login = 'test';
+        $form->password = 'test';
+
+        $form->identityClass = Mock\Identity::class;
+        Mock\Identity::$willFind = $identity = $this->createMock(Mock\Identity::class);
+
+        $identity->expects($this->once())
```

Comment:
It would be better to use `$this->at(int $index)` insted of `$this->once()` to be sure when method called